### PR TITLE
Added AmazonBedrockVectorizer class

### PR DESCRIPTION
**Pull request by dgallitelli**
There was no SentenceVectorizer that leveraged Amazon Bedrock. I've created one.

### CODE REVIEW

File: dsp/modules/sentence_vectorizer.py [hunk 1]
```diff
@@ -247,4 +247,93 @@ def __call__(self, inp_examples: List["Example"]) -> np.ndarray:
...
+        embed_batch_size: int = 128,
+        region_name: str = 'us-west-2',
+        aws_access_key_id: Optional[str] = None,
+        aws_secret_access_key: Optional[str] = None,
+    ):
+        self.model_id = model_id
+        self.embed_batch_size = embed_batch_size
+
+        # Initialize the Bedrock client
+        self.bedrock_client = boto3.client(
```
Comment (arnavsinghvi11) [inline dsp/modules/sentence_vectorizer.py:273]: could this potentially be replaced with the existing [DSPy AWS model integrations](https://github.com/stanfordnlp/dspy/blob/bb97c72da6312600b8ac1a5228f79b07afc427b1/dsp/modules/aws_models.py#L24)? would be neat to tie it in!

File: dsp/modules/sentence_vectorizer.py [hunk 2]
```diff
@@ -247,4 +247,93 @@ def __call__(self, inp_examples: List["Example"]) -> np.ndarray:
...
+                    })
+                else:
+                    raise Exception(f"Model {self.model_id} supports batch size of 1 exclusively.")
+            elif self.model_id.startswith("cohere.embed"):
+                body = json.dumps({
+                    "texts": cur_batch,
+                    "input_type": "search_document"
+                })
+            else:
+                raise Exception("How did you even get here?")
```
Comment (arnavsinghvi11) [inline dsp/modules/sentence_vectorizer.py:313]: would appreciate a more informative exception message :).

File: dsp/modules/sentence_vectorizer.py [hunk 3]
```diff
@@ -1,6 +1,6 @@
 import abc
 from typing import List, Optional
-
+import boto3
```
Comment (arnavsinghvi11) [inline dsp/modules/sentence_vectorizer.py:3]: Update with try/except block like in [aws_providers](https://github.com/stanfordnlp/dspy/blob/bb97c72da6312600b8ac1a5228f79b07afc427b1/dsp/modules/aws_providers.py#L8).

**Comment by arnavsinghvi11** [comment]
Thanks @dgallitelli for the PR! left a few comments with the main change required in wrapping the `import boto3` with a [try/except block](https://github.com/stanfordnlp/dspy/blob/bb97c72da6312600b8ac1a5228f79b07afc427b1/dsp/modules/aws_providers.py#L8) to pass the failing the tests. 

Also, please run  `ruff check . --fix-only` and push to pass the failing ruff test.

File: dsp/modules/sentence_vectorizer.py (hunk 1 above)
Comment (dgallitelli) [reply dsp/modules/sentence_vectorizer.py:273]: I'm not exactly sure what you would like to see here to be honest 🤔

**Comment by dgallitelli** [comment]
Thanks @arnavsinghvi11 for your comments! I've addressed the wrapping of `boto3` and the more explicit error messaging, and pushed the updates to the branch. Also, I've run the `ruff` test, with the following output:

```
warning: The top-level linter settings are deprecated in favour of their counterparts in the `lint` section. Please update the following options in `pyproject.toml`:
  - 'extend-unsafe-fixes' -> 'lint.extend-unsafe-fixes'
warning: Docstring at dspy/teleprompt/mipro_optimizer.py:111:9 contains implicit string concatenation; ignoring...
warning: Docstring at dspy/teleprompt/mipro_optimizer.py:122:9 contains implicit string concatenation; ignoring...
```

This makes me think that all is good so far. 

I'm not sure your point about the existing [DSPy AWS model integrations](https://github.com/stanfordnlp/dspy/blob/bb97c72da6312600b8ac1a5228f79b07afc427b1/dsp/modules/aws_models.py#L24). What would you be looking for here? A new `AWSBedrockEmbedding` class? Just switching the boto3 client to the DSPy AWSProvider? I think that can be a separate Issue and PR.

**Comment by arnavsinghvi11** [comment]
Thanks @dgallitelli . Yes I was curious if the boto3 client could be switched to use the supported DSPy AWS providers. Would be a seamless way to tie in the DSPy library together.

**Pull request closed without merging**